sys/fido2/ctap/ctap.c

```diff
@@ -0,0 +1,1855 @@
+
+    if (ret != CTAP2_OK) {
+        goto done;
+    }
+
+    sz = sizeof(pin_hash_dec);
+    ret = fido2_ctap_crypto_aes_dec(pin_hash_dec, &sz, req->pin_hash_enc,
+                                    sizeof(req->pin_hash_enc), shared_key,
+                                    sizeof(shared_key));
+
+    if (ret != CTAP2_OK) {
+        DEBUG("set pin: error while decrypting pin hash \n");
+        goto done;
+    }
+
+    if (memcmp(pin_hash_dec, _state.pin_hash, sizeof(_state.pin_hash)) != 0) {
+        DEBUG("fido2_ctap: _get_pin_token - invalid pin \n");
+
+        /* reset key agreement key */
+        ret =
+            fido2_ctap_crypto_gen_keypair(&_state.ag_key.pub, _state.ag_key.priv,
+                                          sizeof(_state.ag_key.priv));
+
+        if (ret != CTAP2_OK) {
+            goto done;
+        }
+
+        ret = _decrement_pin_attempts();
+
+        if (ret != CTAP2_OK) {
+            goto done;
+        }
+
+        ret = CTAP2_ERR_PIN_INVALID;
+        goto done;
+    }
+
+    _reset_pin_attempts();
+
+    sz = sizeof(pin_token_enc);
+    ret = fido2_ctap_crypto_aes_enc(pin_token_enc, &sz, _pin_token,
+                                    sizeof(_pin_token), shared_key,
+                                    sizeof(shared_key));
+
+    if (ret != CTAP2_OK) {
+        DEBUG("get pin token: error encrypting pin token \n");
+        goto done;
+    }
+
+    ret = fido2_ctap_cbor_encode_pin_token(pin_token_enc,
+                                           sizeof(pin_token_enc));
+
+done:
+    /* clear key agreement key */
+    memset(&_state.ag_key, 0, sizeof(_state.ag_key));
+    return ret;
+}
+
+static int _save_pin(uint8_t *pin, size_t len)
+{
+    uint8_t buf[SHA256_DIGEST_LENGTH] = { 0 };
+    int ret;
+
+    /* store LEFT(SHA-256(newPin), 16) */
+    ret = fido2_ctap_crypto_sha256(pin, len, buf);
+
+    if (ret != CTAP2_OK) {
+        return ret;
+    }
+
+    memcpy(_state.pin_hash, buf, sizeof(_state.pin_hash));
+    _state.pin_is_set = true;
+
+    _write_state_to_flash(&_state);
+
+    return CTAP2_OK;
+}
+
+bool fido2_ctap_cred_params_supported(uint8_t cred_type, int32_t alg_type)
+{
+    if (cred_type == CTAP_PUB_KEY_CRED_PUB_KEY) {
+        if (alg_type == CTAP_COSE_ALG_ES256) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static inline bool _pin_protocol_supported(uint8_t version)
+{
+    return version == CTAP_PIN_PROT_VER;
+}
+
+bool fido2_ctap_pin_is_set(void)
+{
+    return _state.pin_is_set;
+}
+
+static inline bool _is_locked(void)
+{
+    return _state.rem_pin_att <= 0;
+}
+
+static inline bool _is_boot_locked(void)
+{
+    return _rem_pin_att_boot <= 0;
+}
+
+static int _decrement_pin_attempts(void)
+{
+    if (_state.rem_pin_att > 0) {
+        _state.rem_pin_att--;
+    }
+
+    if (_rem_pin_att_boot > 0) {
+        _rem_pin_att_boot--;
+    }
+
+    _write_state_to_flash(&_state);
```

Comment:
This clears the entire page - is this intentional?

Reply:
Yes since there is nothing else stored in `CTAP_FLASH_STATE_PAGE`. I could slightly improve the function by only writing `sizeof(*state)` instead of `sizeof(page)` many bytes.

Reply:
You could always flip bits from `FLASHPAGE_ERASE_STATE` to `!FLASHPAGE_ERASE_STATE` without clearing the page.

e.g. you could have

```C
uint8_t tries = (1U << PIN_RETRIES) - 1; /* set PIN_RETRIES bits */

bool decrement(uint8_t *tries)
{
    if (*tries == 0) {
        return false;
    }

    *tries &= ~(1 << (ffs(*tries) - 1)); /* clear a single bit */
    return true;
}
```

and then write `tries ^ FLASHPAGE_ERASE_STATE` without clearing the page.